Take image virtualization types into account when selecting proper vi…

cket opened the pull request:
…rtualization type

Issue #1

The order of virtualization type preference was not being preserved, and the image type's allowable virtualization types was not taken into account.

Commit:
Take image virtualization types into account when selecting proper virtualization type

Issue #1

Commit:
Only use _base_image when necessary, and take cgcloud image virtualization type into account.

Commit:
Comments to explain use of lists instead of sets

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -273,14 +273,15 @@ def _populate_security_group( self, group_name ):
-        instance_vtypes = set( ec2_instance_types[ instance_type ].virtualization_types )
+    def __get_virtualization_types( self, instance_type, virtualization_type ):
+        # Use lists and list comprehensions instead of sets. Results should be sorted by order of preference, sets
+        #   lose this order. OrderedSet in bd2k.util.collections does not have an intersection method, which we want.
```
You can use the & operator:

```
>>> from bd2k.util.collections import OrderedSet
>>> OrderedSet('ab') & OrderedSet('bc')
OrderedSet(['b'])
```

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +321,21 @@ def prepare( self, ec2_keypair_globs,
-            image = self._base_image( virtualization_type )
-            log.info( "... found %s.", image.id )
+            for type in virtualization_types:
+                log.info( "Looking up default image for role %s and type %s, ... " % (self.role( ), type) )
```
For log messages, don't use %. Instead of 

```
log.info( "bla %s %s" % ( 1, 2 ) )
```

use

```
log.info( "bla %s %s", 1, 2 )
```

That is shorter and more efficient since the interpolation is only performed if the log level is at or below info.

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -290,6 +291,7 @@ def __get_virtualization_type( self, instance_type, virtualization_type ):
                 raise RuntimeError(
                     'Virtualization type %s not supported by role %s and instance type %s' % (
                         virtualization_type, self.role( ), instance_type ) )
+            instance_type=list(instance_type)
```
What's this for?

hannes-ucsc replied:
It's an unused assignment.

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -273,14 +273,15 @@ def _populate_security_group( self, group_name ):
             if vtypes:
-                # find the preferred vtype, i.e. the one listed first in instance_vtypes
-                virtualization_type = next( vtype for vtype in instance_vtypes if vtype in vtypes )
+                virtualization_type = vtypes
```
Shouldn't the variable name be plural?

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +321,21 @@ def prepare( self, ec2_keypair_globs,
+                log.info( "Looking up default image for role %s and type %s, ... " % (self.role( ), type) )
+                image = self._base_image( type )
+                log.info( "... found %s.", image.id )
+                if image.virtualization_type in virtualization_types:
```
What does _base_image() do if it can't find an image for a given type?

And isn't `self._base_image( type ).virtualization_type == type` always true?

Commit:
Make use of Ordered Sets, improve logging, improve virtualization type searching logic

cket commented:
This commit should address your comments

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -290,7 +290,8 @@ def __get_virtualization_type( self, instance_type, virtualization_type ):
                     'Virtualization type %s not supported by role %s and instance type %s' % (
                         virtualization_type, self.role( ), instance_type ) )
-        return virtualization_type
+            virtualization_types=list([virtualization_type])
```
Calling the list constructor is unnecessary. [] is already a list.

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
+                try:
+                    image = self._base_image( type )
+                except RuntimeError:
+                    log.warning("... did not find a suitable image.")
```
This should be info level since it is expected to happen.

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
 
-        if image.virtualization_type != virtualization_type:
+        if image.virtualization_type not in virtualization_types:
             raise RuntimeError( "Expected virtualization type %s but image only supports %s" % (
```
The wording of this message still assumes that there is only one type.

hannes-ucsc replied:
Also:

```
>>> "%s" % ['hvm','paravirtual']
"['hvm', 'paravirtual']"
```

Can we make that a bit more like an English sentence? It would be overkill just for this but I need it all the time. Maybe we can be inspired by

https://pypi.python.org/pypi/inflect#converting-lists-of-words-to-phrases

It would be fun to write such a method and add it to bd2k-python-lib.

cket replied:
Pull request 
https://github.com/BD2KGenomics/bd2k-python-lib/pull/1

Commit:
Improved logging, fixed virtualization_type list construction

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -290,7 +290,8 @@ def __get_virtualization_type( self, instance_type, virtualization_type ):
                     'Virtualization type %s not supported by role %s and instance type %s' % (
                         virtualization_type, self.role( ), instance_type ) )
-        return virtualization_type
+            virtualization_types=list(virtualization_type)
```
This can't work. `list("hvm")` would yield `[ 'h', 'v', 'm' ]`

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
+                log.info( "Looking up default image for role %s and type %s, ... ", self.role( ), type )
+                try:
+                    image = self._base_image( type )
+                except RuntimeError:
```
Too general.

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
+                    break
+
+        if image is None:
+            raise RuntimeError("Could not find suitable image for role %s", type)
```
Can't use `type` here. Do you mean `self.role()`?

hannes-ucsc commented on core/src/cgcloud/core/box.py:
```diff
@@ -319,18 +320,29 @@ def prepare( self, ec2_keypair_globs,
+                    log.info("... did not find a suitable image.")
+                else:
+                    log.info( "... found %s.", image.id )
+                if image is not None and image.virtualization_type in virtualization_types:
```
The first term of this conjunction can be eliminated by moving the if statement into the else: branch of the try: block above.

hannes-ucsc commented:
There are two obvious problems in this change (see my line notes). This indicates that the code hasn't been tested for all cases. To expedite the process, I'll take it from here.

hannes-ucsc committed:
Fix: direct dependency of agent on bd2k-python-lib is versioned inconsistently with the transitive one through lib

Don't know why pip silently violates the ==1.6.dev1 ref by installing 1.7.dev1, even though it could have satisfied both by using 1.6.dev1 which matches ==1.6.dev1 and >=1.6.dev1.

[52.25.171.223] run: /opt/cgcloudagent/bin/cgcloudagent --init-script --zone us-west-2a --namespace /hannes/ --accounts ubuntu --keypairs hannes@ucsc.edu @@developers --user root --group root --pid-file /var/run/cgcloudagent/cgcloudagent.pid --log-spill /var/log/cgcloudagent.out| gzip -c | base64
[52.25.171.223] out: Traceback (most recent call last):
[52.25.171.223] out:   File "/opt/cgcloudagent/bin/cgcloudagent", line 5, in <module>
[52.25.171.223] out:     from pkg_resources import load_entry_point
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3084, in <module>
[52.25.171.223] out:     @_call_aside
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3070, in _call_aside
[52.25.171.223] out:     f(*args, **kwargs)
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 3097, in _initialize_master_working_set
[52.25.171.223] out:     working_set = WorkingSet._build_master()
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 653, in _build_master
[52.25.171.223] out:     return cls._build_from_requirements(__requires__)
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 666, in _build_from_requirements
[52.25.171.223] out:     dists = ws.resolve(reqs, Environment())
[52.25.171.223] out:   File "/opt/cgcloudagent/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 839, in resolve
[52.25.171.223] out:     raise DistributionNotFound(req, requirers)
[52.25.171.223] out: pkg_resources.DistributionNotFound: The 'bd2k-python-lib==1.6.dev1' distribution was not found and is required by cgcloud-agent
[52.25.171.223] out: H4sIAGH+ulUAAwMAAAAAAAAAAAA=

hannes-ucsc committed:
Make _base_image throw specific exception when suitable image not found; tightened vtype logic
